### PR TITLE
CORE-444 Move methods to basic_api and cache passphrase

### PR DESCRIPTION
**Pull request by ig-garcia**

### CODE REVIEW

File: src/cxx/basic_api.cc [hunk 1]
```diff
@@ -468,6 +468,74 @@ JNIEXPORT void JNICALL Java_foundation_pEp_jniadapter_Engine__1config_1passphras
...
+
+    const stringpair_list_t* _accountswithpassphrases = to_stringpairlist(env, accountswithpassphrases);
+    stringlist_t *_errorAccounts;
+
+    PEP_STATUS status = ::unlock_keys_with_passphrase(session(),_accountswithpassphrases,&_errorAccounts);
+
+    if ((status > PEP_STATUS_OK && status < PEP_UNENCRYPTED) || status < PEP_STATUS_OK || status >= PEP_TRUSTWORD_NOT_FOUND || status == PEP_DECRYPT_SIGNATURE_DOES_NOT_MATCH) {
+        throw_pEp_Exception(env, status);
+        return NULL;
+    }
```
Comment (ghost) [inline src/cxx/basic_api.cc]: We changed the usual contract: `unlock_keys_with_passphrase` can indicate error conditions and _still_ allocate `error_accounts`. In fact the only condition you are guaranteed to have no allocation is a result of PEP_STATUS_OK.

Please see https://github.com/plancksecurity/foundation-planckCOMServerWrapper/blob/CORE-445/CpEpEngine.cpp#L2623 as an example.

File: src/cxx/basic_api.cc [hunk 2]
```diff
@@ -468,6 +468,74 @@ JNIEXPORT void JNICALL Java_foundation_pEp_jniadapter_Engine__1config_1passphras
...
+    const stringpair_list_t* _accountswitholdpassphrases = to_stringpairlist(env, accountswitholdpassphrases);
+    const char *_newpassphrase = to_string(env, newpassphrase);
+    stringlist_t *_errorAccounts;
+
+    PEP_STATUS status = ::manage_passphrase(session(),_accountswitholdpassphrases,_newpassphrase,&_errorAccounts);
+
+    if ((status > PEP_STATUS_OK && status < PEP_UNENCRYPTED) || status < PEP_STATUS_OK || status >= PEP_TRUSTWORD_NOT_FOUND || status == PEP_DECRYPT_SIGNATURE_DOES_NOT_MATCH) {
+        throw_pEp_Exception(env, status);
+        return NULL;
+    }
```
Comment (ghost) [inline src/cxx/basic_api.cc]: Similar to my comment on the call to `unlock_keys_with_passphrase`: The status check is at once too specific/complicated and doesn't take the real behavior into account (which _is_ admittedly a deviation from the usual behavior of core functions).

File: src/cxx/basic_api.cc [hunk 3]
```diff
@@ -468,6 +468,74 @@ JNIEXPORT void JNICALL Java_foundation_pEp_jniadapter_Engine__1config_1passphras
...
+    jobject errorAccounts_ = NULL;
+    if (_errorAccounts) {
+        errorAccounts_ = from_stringlist(env, _errorAccounts);
+    } else {
+        for (const ::stringpair_list_t *curr = _accountswitholdpassphrases; curr != nullptr; curr = curr->next) {
+            char* passphrase = curr->value->value;
+            passphrase_cache.add(passphrase);
+        }
+    }
```
Comment (ghost) [inline src/cxx/basic_api.cc]: Similar to the COM server code, I'd recommend an explicit check here for both PEP_STATUS_OK _and_ `_errorAccounts`.

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Small recap for `unlock_keys_with_passphrase` and `manage_passphrase`:

* If they return `PEP_STATUS_OK` all is good. No `error_accounts` will be set, nothing can be leaked.
* Any other error may have `error_accounts` be set. In that case you can't return early (or throw) without having freed `error_accounts` and/or made sure the caller gets it. Otherwise you're leaking memory and/or don't give the caller important information.

File: src/cxx/basic_api.cc (hunk 1 above)
Comment (ig-garcia) [reply src/cxx/basic_api.cc]: I see, what happens in any other status value? can we not have errors that don't come from the failed verification itself?

Comment (ghost) [reply src/cxx/basic_api.cc]: I wrote a small recap as a general comment to this review.

Comment (ig-garcia) [reply src/cxx/basic_api.cc]: Ok I see, thanks.

**Comment by ig-garcia** [comment]
> Small recap for `unlock_keys_with_passphrase` and `manage_passphrase`:
> 
>     * If they return `PEP_STATUS_OK` all is good. No `error_accounts` will be set, nothing can be leaked.
> 
>     * Any other error may have `error_accounts` be set. In that case you can't return early (or throw) without having freed `error_accounts` and/or made sure the caller gets it. Otherwise you're leaking memory and/or don't give the caller important information.

What happens in this case?

```
if ((status != PEP_STATUS_OK && !_error_accounts) || (status == PEP_STATUS_OK && _error_accounts)) {
        // what happens in these cases? never happens?
}
```

File: src/cxx/basic_api.cc (hunk 1 above)
Comment (ig-garcia) [reply src/cxx/basic_api.cc]: I asked in the recap, what happens in other cases? not possible? out of scope for now?

**Comment by ghost** [comment]
> > Small recap for `unlock_keys_with_passphrase` and `manage_passphrase`:
> > ```
> > * If they return `PEP_STATUS_OK` all is good. No `error_accounts` will be set, nothing can be leaked.
> > 
> > * Any other error may have `error_accounts` be set. In that case you can't return early (or throw) without having freed `error_accounts` and/or made sure the caller gets it. Otherwise you're leaking memory and/or don't give the caller important information.
> > ```
> 
> What happens in this case?
> 
> ```
> if ((status != PEP_STATUS_OK && !_error_accounts) || (status == PEP_STATUS_OK && _error_accounts)) {
>         // what happens in these cases? never happens?
> }
> ```

`status != PEP_STATUS_OK && !_error_accounts` -> Error without `_error_accounts` set. Nothing to free, return the error to the caller, but there are no error accounts. Can actually happen (illegal state in the DB or parameters).

`status == PEP_STATUS_OK && _error_accounts` -> Should not happen, and contradicts itself. If you want to handle it anyways, free the error accounts, but to the caller deliver success. Create a ticket.

**Comment by ig-garcia** [comment]
> > > Small recap for `unlock_keys_with_passphrase` and `manage_passphrase`:
> > > ```
> > > * If they return `PEP_STATUS_OK` all is good. No `error_accounts` will be set, nothing can be leaked.
> > > 
> > > * Any other error may have `error_accounts` be set. In that case you can't return early (or throw) without having freed `error_accounts` and/or made sure the caller gets it. Otherwise you're leaking memory and/or don't give the caller important information.
> > > ```
> > 
> > 
> > What happens in this case?
> > ```
> > if ((status != PEP_STATUS_OK && !_error_accounts) || (status == PEP_STATUS_OK && _error_accounts)) {
> >         // what happens in these cases? never happens?
> > }
> > ```
> 
> `status != PEP_STATUS_OK && !_error_accounts` -> Error without `_error_accounts` set. Nothing to free, return the error to the caller, but there are no error accounts. Can actually happen (illegal state in the DB or parameters).
> 
> `status == PEP_STATUS_OK && _error_accounts` -> Should not happen, and contradicts itself. If you want to handle it anyways, free the error accounts, but to the caller deliver success. Create a ticket.

Ok then I need to throw the exception only in the case of `status != PEP_STATUS_OK && !_error_accounts` -> Error without `_error_accounts`

File: src/cxx/basic_api.cc (hunk 1 above)
Comment (ig-garcia) [reply src/cxx/basic_api.cc]: I updated both methods, still,  if I need to free anything, I don't seem to have this `free_stringlist` method available in JNI...

File: src/cxx/basic_api.cc [hunk 4]
```diff
@@ -468,6 +468,79 @@ JNIEXPORT void JNICALL Java_foundation_pEp_jniadapter_Engine__1config_1passphras
...
+    std::mutex *mutex_local = nullptr;
+    {
+        std::lock_guard<std::mutex> l(global_mutex);
+        pEpLog("called with lock_guard");
+        mutex_local = get_engine_java_object_mutex(env, obj);
+    }
+    std::lock_guard<std::mutex> l(*mutex_local);
+
+    const stringpair_list_t* _accountswithpassphrases = to_stringpairlist(env, accountswithpassphrases);
+    stringlist_t *_errorAccounts;
```
Comment (ghost) [inline src/cxx/basic_api.cc]: Initialize to nullptr, don't rely on core.

File: src/cxx/basic_api.cc [hunk 5]
```diff
@@ -468,6 +468,79 @@ JNIEXPORT void JNICALL Java_foundation_pEp_jniadapter_Engine__1config_1passphras
...
+    {
+        std::lock_guard<std::mutex> l(global_mutex);
+        pEpLog("called with lock_guard");
+        mutex_local = get_engine_java_object_mutex(env, obj);
+    }
+    std::lock_guard<std::mutex> l(*mutex_local);
+
+    stringpair_list_t* _accountswitholdpassphrases = to_stringpairlist(env, accountswitholdpassphrases);
+    const char *_newpassphrase = to_string(env, newpassphrase);
+    stringlist_t *_errorAccounts;
```
Comment (ghost) [inline src/cxx/basic_api.cc]: Initialize to nullptr, don't rely on core.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/cxx/basic_api.cc (hunk 5 above)
Comment (ig-garcia) [reply src/cxx/basic_api.cc]: Done.

File: src/cxx/basic_api.cc (hunk 4 above)
Comment (ig-garcia) [reply src/cxx/basic_api.cc]: Done.

File: src/cxx/basic_api.cc (hunk 3 above)
Comment (ig-garcia) [reply src/cxx/basic_api.cc]: Done.

File: src/cxx/basic_api.cc (hunk 2 above)
Comment (ig-garcia) [reply src/cxx/basic_api.cc]: Done.

File: src/cxx/basic_api.cc (hunk 1 above)
Comment (ig-garcia) [reply src/cxx/basic_api.cc]: Done.

**Comment by ghost** [comment]
@Stereodark @saschabacardit Please take a look from the JNI side of things. The c++/core interaction looks good to me.

**Review by saschabacardit — APPROVED** [review]
I am 99% sure basic_api shouldn't be uploaded as it is the autogen is it not?
Either way, ysl2 looks okay.

**Comment by ig-garcia** [comment]
> I am 99% sure basic_api shouldn't be uploaded as it is the autogen is it not? Either way, ysl2 looks okay.

Yeah, basic_api is in git from ages. So that one is not generated, it is created by hand or generated from the autogen and pasted and commited etc. Then what is generated is the interfaces. Weird yes.

**Review by ghost — APPROVED** [review]

**Pull request merged**